File: internal/types/registry.go
// Package types is the generic engine (Layer 1). It loads the self-describing
// ArtifactType definitions from docs/types/*.md, resolves inheritance, and
// compiles the friendly dialect into checks. The ONLY literal type name in the
// whole codebase is the string "ArtifactType", confined to this package.
package types

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"

	"github.com/PurnaOS/iBuildOS/internal/model"
	"github.com/PurnaOS/iBuildOS/internal/okf"
	"gopkg.in/yaml.v3"
)

// metaType is the only type the engine knows natively.
const metaType = "ArtifactType"

// reserved files in a types dir that are never type definitions.
var reserved = map[string]bool{"index.md": true, "log.md": true}

// validScalarTypes are the field `type:` values the dialect understands.
var validScalarTypes = map[string]bool{
	"": true, "string": true, "number": true, "date": true, "bool": true, "list": true,
}

// FieldSpec is a compiled field check.
type FieldSpec struct {
	Required bool
	OneOf    []string
	Pattern  string
	Type     string // "", string, number, date, bool, list
	Doc      string
	Re       *regexp.Regexp
}

// RelSpec is a compiled relationship check.
type RelSpec struct {
	Target string
	Min    int
	Max    *int // nil = unbounded
	Doc    string
}

// Definition is one ArtifactType document's own (un-inherited) declaration.
type Definition struct {
	Defines     string
	Extends     string
	Abstract    bool
	Description string
	Fields      map[string]FieldSpec
	Rels        map[string]RelSpec
	JSONSchema  *yaml.Node
	Path        string // bundle-relative, for findings
	defLine     int
}

// Resolved is a Definition flattened across its extends chain (child overrides parent).
type Resolved struct {
	Name        string
	Abstract    bool
	Fields      map[string]FieldSpec
	Rels        map[string]RelSpec
	JSONSchemas []*yaml.Node // own + ancestors (applied in addition)
}

// Registry is the compiled type model.
type Registry struct {
	defs     map[string]*Definition
	resolved map[string]*Resolved
	desc     map[string]map[string]bool // type -> set of {self + transitive subtypes}
}

type rawField struct {
	Required bool     `yaml:"required"`
	OneOf    []string `yaml:"one_of"`
	Pattern  string   `yaml:"pattern"`
	Type     string   `yaml:"type"`
	Doc      string   `yaml:"doc"`
}

type rawRel struct {
	Target string `yaml:"target"`
	Min    int    `yaml:"min"`
	Max    *int   `yaml:"max"`
	Doc    string `yaml:"doc"`
}

type rawDef struct {
	Type          string              `yaml:"type"`
	Defines       string              `yaml:"defines"`
	Extends       string              `yaml:"extends"`
	Abstract      bool                `yaml:"abstract"`
	Description   string              `yaml:"description"`
	Fields        map[string]rawField `yaml:"fields"`
	Relationships map[string]rawRel   `yaml:"relationships"`
}

// Load reads every *.md under typesDir, compiles the type model, and meta-
// validates it. Definition problems are emitted as error findings (via c) so
// they fail the build and get annotated; the returned registry is still usable
// for whatever resolved correctly. An unreadable typesDir is a hard error.
func Load(typesDir, bundleDir string, c *model.Collector) (*Registry, error) {
	entries, err := os.ReadDir(typesDir)
	if err != nil {
		return nil, err
	}
	r := &Registry{
		defs:     map[string]*Definition{},
		resolved: map[string]*Resolved{},
		desc:     map[string]map[string]bool{},
	}
	var names []string
	for _, e := range entries {
		names = append(names, e.Name())
	}
	sort.Strings(names)

	for _, name := range names {
		if e := loadOne(r, typesDir, bundleDir, name, c); e != nil {
			continue
		}
	}
	r.checkReferences(c)
	r.buildDescendants()
	return r, nil
}

func loadOne(r *Registry, typesDir, bundleDir, name string, c *model.Collector) error {
	if reserved[name] || filepath.Ext(name) != ".md" {
		return errSkip
	}
	abs := filepath.Join(typesDir, name)
	rel := bundleRel(bundleDir, abs)
	raw, err := os.ReadFile(abs)
	if err != nil {
		return errSkip
	}
	d, err := okf.Parse(abs, raw)
	if err != nil || !d.HasFrontmatter {
		return errSkip // tolerate prose / malformed files in the types dir
	}
	if _, tv, ok := d.Get("type"); !ok || tv.Value != metaType {
		return errSkip // skip non-ArtifactType files (e.g. overview.md type: Reference)
	}

	var rd rawDef
	if err := d.Map.Decode(&rd); err != nil {
		c.Errf(rel, d.FrontStartLine(), "types.badMeta", "cannot read type definition: %v", err)
		return errSkip
	}
	if rd.Defines == "" {
		c.Errf(rel, d.FrontStartLine(), "types.badMeta", "ArtifactType definition is missing a `defines` name")
		return errSkip
	}
	if existing, dup := r.defs[rd.Defines]; dup {
		c.Errf(rel, d.FrontStartLine(), "types.duplicate",
			"type %q is already defined in %s", rd.Defines, existing.Path)
		return errSkip
	}

	def := &Definition{
		Defines:     rd.Defines,
		Extends:     rd.Extends,
		Abstract:    rd.Abstract,
		Description: rd.Description,
		Fields:      map[string]FieldSpec{},
		Rels:        map[string]RelSpec{},
		Path:        rel,
		defLine:     d.FrontStartLine(),
	}
	// Capture the json_schema block straight from the document mapping (more
	// reliable than struct-decoding a *yaml.Node).
	if _, js, ok := d.Get("json_schema"); ok {
		def.JSONSchema = js
	}
	for fname, rf := range rd.Fields {
		fs := FieldSpec{Required: rf.Required, OneOf: rf.OneOf, Pattern: rf.Pattern, Type: rf.Type, Doc: rf.Doc}
		if !validScalarTypes[rf.Type] {
			c.Errf(rel, d.FrontStartLine(), "types.badMeta",
				"field %q in type %q has unknown type %q (want string|number|date|bool|list)", fname, rd.Defines, rf.Type)
		}
		if rf.Pattern != "" {
			re, perr := compilePattern(rf.Pattern)
			if perr != nil {
				c.Errf(rel, d.FrontStartLine(), "types.badPattern",
					"field %q in type %q has an invalid pattern %q: %v", fname, rd.Defines, rf.Pattern, perr)
			} else {
				fs.Re = re
			}
		}
		def.Fields[fname] = fs
	}
	for rname, rr := range rd.Relationships {
		if rr.Target == "" {
			c.Errf(rel, d.FrontStartLine(), "types.badMeta",
				"relationship %q in type %q is missing a `target`", rname, rd.Defines)
		}
		def.Rels[rname] = RelSpec{Target: rr.Target, Min: rr.Min, Max: rr.Max, Doc: rr.Doc}
	}
	r.defs[rd.Defines] = def
	return nil
}

// checkReferences validates that every extends / relationship target names a
// defined type, and that the extends graph is acyclic.
func (r *Registry) checkReferences(c *model.Collector) {
	for _, name := range r.sortedDefNames() {
		def := r.defs[name]
		if def.Extends != "" {
			if _, ok := r.defs[def.Extends]; !ok {
				c.Errf(def.Path, def.defLine, "types.unknownExtends",
					"type %q extends unknown type %q", name, def.Extends)
			}
		}
		for _, rname := range sortedRelNames(def.Rels) {
			tgt := def.Rels[rname].Target
			if tgt != "" {
				if _, ok := r.defs[tgt]; !ok {
					c.Errf(def.Path, def.defLine, "types.unknownTarget",
						"relationship %q in type %q targets unknown type %q", rname, name, tgt)
				}
			}
		}
	}
	// cycle detection over extends
	const white, gray, black = 0, 1, 2
	color := map[string]int{}
	var visit func(n string) bool
	visit = func(n string) bool {
		color[n] = gray
		if def, ok := r.defs[n]; ok && def.Extends != "" {
			switch color[def.Extends] {
			case gray:
				c.Errf(def.Path, def.defLine, "types.cycle", "type %q is part of an extends cycle", n)
				return true
			case white:
				if visit(def.Extends) {
					return true
				}
			}
		}
		color[n] = black
		return false
	}
	for _, name := range r.sortedDefNames() {
		if color[name] == white {
			if visit(name) {
				break
			}
		}
	}
}

// ancestors returns the extends chain of name (self first, then parents),
// guarding against cycles.
func (r *Registry) ancestors(name string) []string {
	var chain []string
	seen := map[string]bool{}
	for cur := name; cur != ""; {
		if seen[cur] {
			break
		}
		seen[cur] = true
		chain = append(chain, cur)
		def, ok := r.defs[cur]
		if !ok {
			break
		}
		cur = def.Extends
	}
	return chain
}

func (r *Registry) buildDescendants() {
	for name := range r.defs {
		for _, anc := range r.ancestors(name) {
			if r.desc[anc] == nil {
				r.desc[anc] = map[string]bool{}
			}
			r.desc[anc][name] = true
		}
	}
}

// Resolve flattens a type across its extends chain (child overrides parent).
// Returns false for an unknown type.
func (r *Registry) Resolve(name string) (*Resolved, bool) {
	if res, ok := r.resolved[name]; ok {
		return res, true
	}
	def, ok := r.defs[name]
	if !ok {
		return nil, false
	}
	res := &Resolved{Name: name, Abstract: def.Abstract, Fields: map[string]FieldSpec{}, Rels: map[string]RelSpec{}}
	chain := r.ancestors(name)
	// apply parents first so children override
	for i := len(chain) - 1; i >= 0; i-- {
		d, ok := r.defs[chain[i]]
		if !ok {
			continue
		}
		for k, v := range d.Fields {
			res.Fields[k] = v
		}
		for k, v := range d.Rels {
			res.Rels[k] = v
		}
		if d.JSONSchema != nil {
			res.JSONSchemas = append(res.JSONSchemas, d.JSONSchema)
		}
	}
	r.resolved[name] = res
	return res, true
}

// Satisfies reports whether a document of type docType satisfies a relationship
// whose target is target — true iff docType is target or transitively extends it.
func (r *Registry) Satisfies(docType, target string) bool {
	return r.desc[target] != nil && r.desc[target][docType]
}

// Has reports whether a type is defined.
func (r *Registry) Has(name string) bool {
	_, ok := r.defs[name]
	return ok
}

// ConcreteSubtypes returns the non-abstract types that are target or extend it,
// sorted — used to suggest replacements when an abstract type is used directly.
func (r *Registry) ConcreteSubtypes(name string) []string {
	var out []string
	for sub := range r.desc[name] {
		if d, ok := r.defs[sub]; ok && !d.Abstract {
			out = append(out, sub)
		}
	}
	sort.Strings(out)
	return out
}

// RelTarget returns the target type of the first defined relationship with the
// given name (relationship names are shared across types, e.g. implements ->
// Requirement). Returns "" if no type declares it.
func (r *Registry) RelTarget(relName string) string {
	for _, name := range r.sortedDefNames() {
		if rel, ok := r.defs[name].Rels[relName]; ok {
			return rel.Target
		}
	}
	return ""
}

// DefNames returns every defined type name (concrete and abstract), sorted. The
// graph export uses it to project the registry as data.
func (r *Registry) DefNames() []string { return r.sortedDefNames() }

// Ancestors returns name's extends chain, self first then parents (cycle-safe).
func (r *Registry) Ancestors(name string) []string { return r.ancestors(name) }

// IsAbstract reports whether a defined type is abstract.
func (r *Registry) IsAbstract(name string) bool {
	d, ok := r.defs[name]
	return ok && d.Abstract
}

// Description returns the one-line description of a defined type, or "".
func (r *Registry) Description(name string) string {
	if d, ok := r.defs[name]; ok {
		return d.Description
	}
	return ""
}

// Extends returns the immediate parent type of name, or "".
func (r *Registry) Extends(name string) string {
	if d, ok := r.defs[name]; ok {
		return d.Extends
	}
	return ""
}

func (r *Registry) sortedDefNames() []string {
	out := make([]string, 0, len(r.defs))
	for n := range r.defs {
		out = append(out, n)
	}
	sort.Strings(out)
	return out
}

func sortedRelNames(m map[string]RelSpec) []string {
	out := make([]string, 0, len(m))
	for n := range m {
		out = append(out, n)
	}
	sort.Strings(out)
	return out
}

var errSkip = skipErr{}

type skipErr struct{}

func (skipErr) Error() string { return "skip" }

func bundleRel(bundleDir, abs string) string {
	rel, err := filepath.Rel(bundleDir, abs)
	if err != nil {
		return filepath.ToSlash(abs)
	}
	return filepath.ToSlash(rel)
}

File: src/core/types/registry.ts
// The generic engine (Layer 1). Loads the self-describing ArtifactType
// definitions from docs/types/*.md, resolves inheritance, and compiles the
// friendly dialect into checks. Port of Go internal/types/registry.go.
//
// The ONLY literal type name in the whole engine is the string "ArtifactType",
// confined to this file (scripts/check-literals.ts enforces it).
import { readdirSync, readFileSync } from "node:fs";
import { join, relative, extname } from "node:path";
import { Collector } from "../model/model.ts";
import { parse as parseOkf } from "../okf/frontmatter.ts";
import { compilePattern } from "./pattern.ts";

// metaType is the only type the engine knows natively.
const metaType = "ArtifactType";

// reserved files in a types dir that are never type definitions.
const reserved = new Set(["index.md", "log.md"]);

// validScalarTypes are the field `type:` values the dialect understands.
const validScalarTypes = new Set(["", "string", "number", "date", "bool", "list"]);

export interface FieldSpec {
  required: boolean;
  oneOf: string[];
  pattern: string;
  type: string; // "" | string | number | date | bool | list
  doc: string;
  re: RegExp | null;
}

export interface RelSpec {
  target: string;
  min: number;
  max: number | null; // null = unbounded
  doc: string;
}

interface Definition {
  defines: string;
  extends: string;
  abstract: boolean;
  description: string;
  fields: Map<string, FieldSpec>;
  rels: Map<string, RelSpec>;
  jsonSchema: unknown | null; // plain JS object for ajv, null if absent
  path: string; // bundle-relative, for findings
  defLine: number;
}

// Resolved is a Definition flattened across its extends chain (child overrides parent).
export interface Resolved {
  name: string;
  abstract: boolean;
  fields: Map<string, FieldSpec>;
  rels: Map<string, RelSpec>;
  jsonSchemas: unknown[]; // own + ancestors (applied in addition)
}

function asStr(v: unknown): string {
  return typeof v === "string" ? v : "";
}
function asStrList(v: unknown): string[] {
  return Array.isArray(v) ? v.map(String) : [];
}
function asObj(v: unknown): Record<string, unknown> {
  return v && typeof v === "object" && !Array.isArray(v) ? (v as Record<string, unknown>) : {};
}

export class Registry {
  private defs = new Map<string, Definition>();
  private resolvedCache = new Map<string, Resolved>();
  private desc = new Map<string, Set<string>>(); // type -> {self + transitive subtypes}

  // load reads every *.md under typesDir, compiles the type model, and
  // meta-validates it. Definition problems become error findings (via c). An
  // unreadable typesDir throws.
  static load(typesDir: string, bundleDir: string, c: Collector): Registry {
    const r = new Registry();
    const names = readdirSync(typesDir).sort();
    for (const name of names) r.loadOne(typesDir, bundleDir, name, c);
    r.checkReferences(c);
    r.buildDescendants();
    return r;
  }

  private loadOne(typesDir: string, bundleDir: string, name: string, c: Collector): void {
    if (reserved.has(name) || extname(name) !== ".md") return;
    const abs = join(typesDir, name);
    const rel = bundleRel(bundleDir, abs);
    let raw: string;
    try {
      raw = readFileSync(abs, "utf8");
    } catch {
      return;
    }
    const [d, err] = parseOkf(abs, raw);
    if (err || !d.hasFrontmatter) return; // tolerate prose / malformed files
    const tv = d.get("type");
    if (!tv || asStr((tv.valNode as { value?: unknown }).value) !== metaType) return; // skip non-meta files

    const rd = d.toJS();
    const defines = asStr(rd.defines);
    if (defines === "") {
      c.errf(rel, d.frontStartLine(), "types.badMeta", "ArtifactType definition is missing a `defines` name");
      return;
    }
    const existing = this.defs.get(defines);
    if (existing) {
      c.errf(rel, d.frontStartLine(), "types.duplicate", `type ${q(defines)} is already defined in ${existing.path}`);
      return;
    }

    const def: Definition = {
      defines,
      extends: asStr(rd.extends),
      abstract: rd.abstract === true,
      description: asStr(rd.description),
      fields: new Map(),
      rels: new Map(),
      jsonSchema: rd.json_schema ?? null,
      path: rel,
      defLine: d.frontStartLine(),
    };

    const fields = asObj(rd.fields);
    for (const fname of Object.keys(fields)) {
      const rf = asObj(fields[fname]);
      const type = asStr(rf.type);
      const fs: FieldSpec = {
        required: rf.required === true,
        oneOf: asStrList(rf.one_of),
        pattern: asStr(rf.pattern),
        type,
        doc: asStr(rf.doc),
        re: null,
      };
      if (!validScalarTypes.has(type)) {
        c.errf(rel, def.defLine, "types.badMeta",
          `field ${q(fname)} in type ${q(defines)} has unknown type ${q(type)} (want string|number|date|bool|list)`);
      }
      if (fs.pattern !== "") {
        try {
          fs.re = compilePattern(fs.pattern);
        } catch (e) {
          c.errf(rel, def.defLine, "types.badPattern",
            `field ${q(fname)} in type ${q(defines)} has an invalid pattern ${q(fs.pattern)}: ${(e as Error).message}`);
        }
      }
      def.fields.set(fname, fs);
    }

    const rels = asObj(rd.relationships);
    for (const rname of Object.keys(rels)) {
      const rr = asObj(rels[rname]);
      const target = asStr(rr.target);
      if (target === "") {
        c.errf(rel, def.defLine, "types.badMeta",
          `relationship ${q(rname)} in type ${q(defines)} is missing a \`target\``);
      }
      def.rels.set(rname, {
        target,
        min: typeof rr.min === "number" ? rr.min : 0,
        max: typeof rr.max === "number" ? rr.max : null,
        doc: asStr(rr.doc),
      });
    }

    this.defs.set(defines, def);
  }

  // checkReferences validates that every extends / relationship target names a
  // defined type, and that the extends graph is acyclic.
  private checkReferences(c: Collector): void {
    for (const name of this.sortedDefNames()) {
      const def = this.defs.get(name)!;
      if (def.extends !== "" && !this.defs.has(def.extends)) {
        c.errf(def.path, def.defLine, "types.unknownExtends",
          `type ${q(name)} extends unknown type ${q(def.extends)}`);
      }
      for (const rname of [...def.rels.keys()].sort()) {
        const tgt = def.rels.get(rname)!.target;
        if (tgt !== "" && !this.defs.has(tgt)) {
          c.errf(def.path, def.defLine, "types.unknownTarget",
            `relationship ${q(rname)} in type ${q(name)} targets unknown type ${q(tgt)}`);
        }
      }
    }
    // cycle detection over extends (white/gray/black DFS)
    const White = 0, Gray = 1, Black = 2;
    const color = new Map<string, number>();
    const visit = (n: string): boolean => {
      color.set(n, Gray);
      const def = this.defs.get(n);
      if (def && def.extends !== "") {
        switch (color.get(def.extends) ?? White) {
          case Gray:
            c.errf(def.path, def.defLine, "types.cycle", `type ${q(n)} is part of an extends cycle`);
            return true;
          case White:
            if (visit(def.extends)) return true;
        }
      }
      color.set(n, Black);
      return false;
    };
    for (const name of this.sortedDefNames()) {
      if ((color.get(name) ?? White) === White) {
        if (visit(name)) break;
      }
    }
  }

  // ancestors returns the extends chain of name (self first, then parents),
  // guarding against cycles.
  private ancestorChain(name: string): string[] {
    const chain: string[] = [];
    const seen = new Set<string>();
    let cur = name;
    while (cur !== "") {
      if (seen.has(cur)) break;
      seen.add(cur);
      chain.push(cur);
      const def = this.defs.get(cur);
      if (!def) break;
      cur = def.extends;
    }
    return chain;
  }

  private buildDescendants(): void {
    for (const name of this.defs.keys()) {
      for (const anc of this.ancestorChain(name)) {
        let set = this.desc.get(anc);
        if (!set) {
          set = new Set();
          this.desc.set(anc, set);
        }
        set.add(name);
      }
    }
  }

  // resolve flattens a type across its extends chain (child overrides parent).
  resolve(name: string): Resolved | undefined {
    const cached = this.resolvedCache.get(name);
    if (cached) return cached;
    const def = this.defs.get(name);
    if (!def) return undefined;
    const res: Resolved = { name, abstract: def.abstract, fields: new Map(), rels: new Map(), jsonSchemas: [] };
    const chain = this.ancestorChain(name);
    // apply parents first so children override
    for (let i = chain.length - 1; i >= 0; i--) {
      const d = this.defs.get(chain[i]!);
      if (!d) continue;
      for (const [k, v] of d.fields) res.fields.set(k, v);
      for (const [k, v] of d.rels) res.rels.set(k, v);
      if (d.jsonSchema != null) res.jsonSchemas.push(d.jsonSchema);
    }
    this.resolvedCache.set(name, res);
    return res;
  }

  // satisfies reports whether docType is-or-transitively-extends target.
  satisfies(docType: string, target: string): boolean {
    return this.desc.get(target)?.has(docType) ?? false;
  }

  has(name: string): boolean {
    return this.defs.has(name);
  }

  // concreteSubtypes returns the non-abstract types that are target or extend it,
  // sorted — used to suggest replacements when an abstract type is used directly.
  concreteSubtypes(name: string): string[] {
    const out: string[] = [];
    for (const sub of this.desc.get(name) ?? []) {
      const d = this.defs.get(sub);
      if (d && !d.abstract) out.push(sub);
    }
    return out.sort();
  }

  // relTargets returns every distinct target type declared for a relationship
  // name across all types, sorted. A capability predicate must accept is-or-
  // extends ANY of these, never just one — so this is order-independent.
  relTargets(relName: string): string[] {
    const set = new Set<string>();
    for (const name of this.sortedDefNames()) {
      const rel = this.defs.get(name)!.rels.get(relName);
      if (rel && rel.target !== "") set.add(rel.target);
    }
    return [...set].sort();
  }

  // satisfiesAny reports whether docType is-or-extends any of the targets — the
  // order-independent capability predicate built on the runtime type graph.
  satisfiesAny(docType: string, targets: string[]): boolean {
    return targets.some((t) => this.satisfies(docType, t));
  }

  defNames(): string[] {
    return this.sortedDefNames();
  }

  ancestors(name: string): string[] {
    return this.ancestorChain(name);
  }

  isAbstract(name: string): boolean {
    return this.defs.get(name)?.abstract ?? false;
  }

  description(name: string): string {
    return this.defs.get(name)?.description ?? "";
  }

  extendsOf(name: string): string {
    return this.defs.get(name)?.extends ?? "";
  }

  private sortedDefNames(): string[] {
    return [...this.defs.keys()].sort();
  }
}

function q(s: string): string {
  return JSON.stringify(s);
}

function bundleRel(bundleDir: string, abs: string): string {
  const rel = relative(bundleDir, abs);
  return (rel === "" ? abs : rel).replaceAll("\\", "/");
}
